### PR TITLE
Power support

### PR DESCRIPTION
**Pull request by npanpaliya**

### CODE REVIEW

**Comment by bazel-io** [comment]
Can one of the admins verify this patch?

File: src/test/shell/bazel/bazel_toolchain_test.sh [hunk 1]
```diff
@@ -22,8 +22,8 @@ CURRENT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
 source "${CURRENT_DIR}/../integration_test_setup.sh" \
   || { echo "integration_test_setup.sh not found!" >&2; exit 1; }
 
-if [ "${PLATFORM-}" = "darwin" ]; then
-  echo "Skipping test: linaro toolchain is not supported on darwin host."
+if [ "${PLATFORM-}" = "darwin" -o "${MACHINE_TYPE}" = "ppc64le" ]; then
+  echo "Skipping test: linaro toolchain is not supported on darwin host or ppc64le Linux."
```
Comment (damienmg) [inline src/test/shell/bazel/bazel_toolchain_test.sh]: Can we change that test to check that platform is linux instead?

**Comment by damienmg** [comment]
We will test this change after rebase.

**Comment by bbelgodere** [comment]
I just tried bazel-0.4.1-dist.zip on a POWER8 machine with RHEL 7.2 and still failing.

### Environment
```
uname -a
Linux dccpc214.pok.ibm.com 3.10.0-327.el7.ppc64le #1 SMP Thu Oct 29 17:31:13 EDT 2015 ppc64le ppc64le ppc64le GNU/Linux

java -version
openjdk version "1.8.0_65"
OpenJDK Runtime Environment (build 1.8.0_65-b17)
OpenJDK 64-Bit Server VM (build 25.65-b01, mixed mode)
```

my home directory is mounted via NFS so need to set TEST_TMPDIR to avoid issues
```
export TEST_TMPDIR=/tmp/bazel_temp 
wget https://github.com/bazelbuild/bazel/releases/download/0.4.1/bazel-0.4.1-dist.zip
mkdir bazel-0.4.1 
unzip bazel-0.4.1-dist.zip -d bazel-0.4.1
cd bazel-0.4.1
  Building Bazel from scratch.third_party/protobuf/3.0.0/protoc-3.0.0-linux-x86_32.exe -Isrc/main/protobuf/ -Isrc/main/java/com/google/devtools/build/lib/buildeventstream/proto/ --java_out=/tmp/bazel_bmSQ4F0F/src --plugin=protoc-gen-grpc=third_party/grpc/protoc-gen-grpc-java-0.15.0-linux-x86_32.exe --grpc_out=/tmp/bazel_bmSQ4F0F/src src/main/java/com/google/devtools/build/lib/buildeventstream/proto/build_event_stream.proto
scripts/bootstrap/buildenv.sh: line 152: third_party/protobuf/3.0.0/protoc-3.0.0-linux-x86_32.exe: cannot execute binary file
```

**Comment by damienmg** [comment]
Jenkins test this please

File: src/test/shell/bazel/bazel_toolchain_test.sh (hunk 1 above)
Comment (npanpaliya) [inline src/test/shell/bazel/bazel_toolchain_test.sh]: After rebase, the test case started giving timeout error, even with checking linux instead of ppc64le or without any change i.e. just checking darwin. Is this a known problem? If it is, I'll undo this change and keep that original change of darwin.

Comment (damienmg) [inline src/test/shell/bazel/bazel_toolchain_test.sh]: Our CI test is green so there is no issue. If the test is timing out on  supported platform it might be because the platform is overloaded while running the test.

Comment (damienmg) [inline src/test/shell/bazel/bazel_toolchain_test.sh]: Yes I mean this test is made to work only on Linux x86_64 so it's better to test that we are in the good architecture rather than list all the platform it cannot work on.

**Comment by npanpaliya** [comment]
Incorporated the requested change. Please review and let me know if any changes are required.

File: src/main/java/com/google/devtools/build/lib/rules/java/DeployArchiveBuilder.java [hunk 2]
```diff
@@ -258,7 +258,7 @@ public void build() throws InterruptedException {
         outputJar, javaStartClass, deployManifestLines, buildInfoArtifacts, classpathResources,
         runtimeClasspath, includeBuildData, compression, launcher);
 
-    List<String> jvmArgs = ImmutableList.of("-client", SINGLEJAR_MAX_MEMORY);
```
Comment (damienmg) [inline src/main/java/com/google/devtools/build/lib/rules/java/DeployArchiveBuilder.java]: /cc @philwo: why do we still have -client flag on Java on some machine, is there case it would be needed?

**Comment by damienmg** [comment]
I just have one nit left and pinged another developer to get a precision to know if it is safe to remove "-client"

**Comment by damienmg** [comment]
LGTM, I am merging it, if Klaus has no further objection the patch will be merged (Klaus away today so it will not be merged before monday).

A last test from jenkins too: test this please.

Thank you very much for your contribution and my apologies for the initial delay.

**Comment by npanpaliya** [comment]
@damienmg Thank you very much for your review and approval.

File: src/tools/singlejar/diag.h [hunk 3]
```diff
@@ -19,7 +19,7 @@
  * Various useful diagnostics functions from Linux err.h file, wrapped
  * for portability.
  */
-#if defined(__APPLE__) || defined(__linux)
+#if defined(__APPLE__) || defined(__linux__)
```
Comment (aehlig) [inline src/tools/singlejar/diag.h]: How is this change related to supporting power? Also, are we sure the case `defined(__linux)` was wrong in the first place and is not needed any more?

Comment (npanpaliya) [inline src/tools/singlejar/diag.h]: On Power, when we compile a program using -std=c++11, then __linux gets undefined, which otherwise is defined. However, on x86, -std=c++11 does not make any difference. I verified on x86 too, and __linux__ is defined there as well.
On Power:
```
$ cpp -dM foo.c | grep linux
#define __linux 1
#define __linux__ 1
#define __gnu_linux__ 1
#define linux 1
$ cpp -dM -std=c++11 foo.c | grep linux
#define __linux__ 1
#define __gnu_linux__ 1
```

On x86:
```
$ cpp -dM foo.cc | grep linux
#define __linux 1
#define __linux__ 1
#define __gnu_linux__ 1
#define linux 1
~$ cpp -dM -std=c++11 foo.cc | grep linux
#define __linux 1
#define __linux__ 1
#define __gnu_linux__ 1

```

Comment (aehlig) [inline src/tools/singlejar/diag.h]: Thanks for checking. LGTM.

Comment (npanpaliya) [inline src/tools/singlejar/diag.h]: Thank you.

**Comment by npanpaliya** [comment]
Could somebody point me to what is wrong with the test being failing on darwin platform as per CI build? Ideally the test should be skipped on darwin platform.

File: src/test/shell/bazel/bazel_toolchain_test.sh [hunk 4]
```diff
@@ -22,12 +22,7 @@ CURRENT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
 source "${CURRENT_DIR}/../integration_test_setup.sh" \
   || { echo "integration_test_setup.sh not found!" >&2; exit 1; }
 
-expected_platform=0
-if [ "${PLATFORM-}" = "linux" && "${MACHINE_TYPE}" = "x86_64" ]; then
-  expected_platform = 1
-fi
-
-if [$expected_platform -eq 0]
+if ![ "${PLATFORM-}" = "linux" && "${MACHINE_TYPE}" = "x86_64" ]; then
```
Comment (damienmg) [inline src/test/shell/bazel/bazel_toolchain_test.sh]: Add a space between "!" and "["

**Comment by damienmg** [comment]
Jenkins test this please

**Comment by damienmg** [comment]
Oops I should not have done that, the change is still missing

**Comment by npanpaliya** [comment]
Thanks @damienmg. I've pushed that change.

**Comment by damienmg** [comment]
Jenkins test this please
